chain-factory/src/factory.rs

```diff
@@ -30,29 +36,104 @@ pub trait FactoryModule {
         min_validators: usize,
         max_validators: usize,
         min_stake: BigUint,
+        chain_name: ManagedBuffer,
         additional_stake_required: MultiValueEncoded<StakeMultiArg<Self::Api>>,
     ) {
         let payment_amount = self.call_value().egld_value().clone_value();
         let deploy_cost = self.deploy_cost().get();
         require!(payment_amount == deploy_cost, "Invalid payment amount");
 
-        let caller = self.blockchain().get_caller();
         let source_address = self.chain_config_template().get();
         let metadata =
             CodeMetadata::PAYABLE_BY_SC | CodeMetadata::UPGRADEABLE | CodeMetadata::READABLE;
+        let args = self.get_deploy_chain_config_args(
```

Comment:
The `chainId` (preferred or random) is a complex thing because you have one `chain-factory` in each shard and you have to make sure is unique across all shards.

Reply:
maybe keep one factory contract on the main chain that has all the chainIds

Reply:
I created a storage mapper designed for all the chain_ids out there, we can use that on the MainChain to keep track of all the others chain_ids